src/raster/types.rs

```diff
@@ -1,41 +1,482 @@
 
+/// Provides ergonomic access to functions describing [`GDALDataType`] ordinals.
+///
+/// A [`GDALDataType`] indicates the primitive storage value of a cell/pixel in a [`RasterBand`][crate::raster::RasterBand].
+///
+/// # Example
+/// ```rust, no_run
+/// use gdal::raster::{GdalType, GdalTypeDescriptor};
+/// let td = <u32>::descriptor();
+/// println!("{} is {} and uses {} bits.",
+///     td.name(),
+///     if td.is_signed() { "signed" } else { "unsigned" },
+///     td.bits()
+/// );
+/// ```
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+pub struct GdalTypeDescriptor(GDALDataType::Type);
+
+impl GdalTypeDescriptor {
+    /// Find `GdalTypeDescriptor` by name, as would be returned by [`name`][Self::name].
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use gdal::raster::{GdalType, GdalTypeDescriptor};
+    /// assert_eq!(GdalTypeDescriptor::from_name("UInt16").unwrap(), <u16>::descriptor())
+    /// ```
+    #[allow(non_snake_case)]
+    pub fn from_name(name: &str) -> Result<Self> {
+        let c_name = CString::new(name.to_owned())?;
+        let gdal_type = unsafe { GDALGetDataTypeByName(c_name.as_ptr()) };
+        match gdal_type {
+            GDALDataType::GDT_Unknown => Err(GdalError::BadArgument(format!(
+                "unable to find datatype with name '{}'",
+                name
+            ))),
+            _ => gdal_type.try_into(),
+        }
+    }
+
+    /// Finds the smallest data type able to support the provided value.
+    ///
+    /// See [`GDALFindDataTypeForValue`](https://gdal.org/api/raster_c_api.html#_CPPv424GDALFindDataTypeForValuedi)
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use gdal::raster::{GdalType, GdalTypeDescriptor};
+    /// assert_eq!(GdalTypeDescriptor::for_value(0), <u8>::descriptor());
+    /// assert_eq!(GdalTypeDescriptor::for_value(256), <u16>::descriptor());
+    /// assert_eq!(GdalTypeDescriptor::for_value(-1), <i16>::descriptor());
+    /// assert_eq!(GdalTypeDescriptor::for_value(<u16>::MAX as f64 * -2.0), <i32>::descriptor());
+    /// ```
+    pub fn for_value<N: GdalType + Into<f64>>(value: N) -> Self {
+        let gdal_type = unsafe { GDALFindDataTypeForValue(value.into(), 0) };
+        GdalTypeDescriptor(gdal_type)
+    }
+
+    /// Get the [`GDALDataType`] ordinal value
+    pub fn gdal_type(&self) -> GDALDataType::Type {
+        self.0
+    }
+
+    /// Get the name of the [`GDALDataType`].
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use gdal::raster::GdalType;
+    /// assert_eq!(<u16>::descriptor().name(), "UInt16");
+    /// ```
+    pub fn name(&self) -> String {
+        let c_str = unsafe { GDALGetDataTypeName(self.gdal_type()) };
+        if c_str.is_null() {
+            // This case shouldn't happen, because `self` only exists for valid
+            // GDALDataType ordinals.
+            panic!("{}", _last_null_pointer_err("GDALGetDataTypeName"));
```

Comment:
`GDALGetDataTypeName` doesn't set an error.

Reply:
Thanks. I need to do a better job of reading the C docs on return type semantics.